cpp/src/copying/copy.cu

```diff
@@ -183,55 +159,89 @@ class logical_not {
-      table_view{std::vector<column_view>{scatter_src_rhs->get_column(0).view()}},
-      scatter_map_rhs.begin(),
-      scatter_map_end,
-      table_view{std::vector<column_view>{lhs}},
-      false,
-      stream,
-      mr);
-
-    return std::move(result->release()[0]);
-  }
+  auto scatter_map = rmm::device_uvector<size_type>{static_cast<std::size_t>(size), stream};
+  auto const scatter_map_end = thrust::copy_if(rmm::exec_policy(stream),
+                                               thrust::make_counting_iterator(size_type{0}),
+                                               thrust::make_counting_iterator(size_type{size}),
+                                               scatter_map.begin(),
+                                               is_left);
+
+  auto const scatter_src_lhs = cudf::detail::gather(table_view{std::vector<column_view>{lhs}},
+                                                    scatter_map.begin(),
+                                                    scatter_map_end,
+                                                    out_of_bounds_policy::DONT_CHECK,
+                                                    stream);
+
+  auto result = cudf::detail::scatter(
+    table_view{std::vector<column_view>{scatter_src_lhs->get_column(0).view()}},
+    scatter_map.begin(),
+    scatter_map_end,
+    table_view{std::vector<column_view>{rhs}},
+    false,
+    stream,
+    mr);
+
+  return std::move(result->release()[0]);
+}
 
-  // Bail out for Scalars.
-  // For nested types types, scatter/gather based copy_if_else() is not currently supported
-  // if either `lhs` or `rhs` is a scalar, partially because:
-  //   1. Struct scalars are not yet available.
-  //   2. List scalars do not yet support explosion to a full column.
-  CUDF_FAIL("Scalars of nested types are not currently supported!");
-  (void)lhs;
-  (void)rhs;
-  (void)size;
-  (void)is_left;
-  (void)stream;
-  (void)mr;
+template <typename Filter>
+std::unique_ptr<column> scatter_gather_based_if_else(cudf::scalar const& lhs,
+                                                     cudf::column_view const& rhs,
+                                                     size_type size,
+                                                     Filter is_left,
+                                                     rmm::cuda_stream_view stream,
+                                                     rmm::mr::device_memory_resource* mr)
+{
+  auto scatter_map = rmm::device_uvector<size_type>{static_cast<std::size_t>(size), stream};
+  auto const scatter_map_end = thrust::copy_if(rmm::exec_policy(stream),
+                                               thrust::make_counting_iterator(size_type{0}),
+                                               thrust::make_counting_iterator(size_type{size}),
+                                               scatter_map.begin(),
+                                               is_left);
+
+  auto const scatter_map_size  = scatter_map_end - scatter_map.begin();
```

Comment:
Nitpick: I have been corrected in the past to use `std::distance()` instead of explicit subtraction, for readability:
```suggestion
  auto const scatter_map_size  = std::distance(scatter_map.begin(), scatter_map_end);
```